### PR TITLE
Introduce protobuf libraries

### PR DESCRIPTION
**Pull request by mara214**
Introduces:

- The `lib` package types (and thus adjusts the manifest format)
- Scoped and nested dependency resolution (installing libraries needed for api packages you consume)

### CODE REVIEW

File: src/package.rs [hunk 1]
```diff
@@ -57,28 +65,75 @@ impl PackageStore {
...
+                '┗'
+            } else {
+                '┣'
+            };
+
+            tracing::info!(
+                "   {tree_char} installed {}@{}",
+                dependency.manifest.name,
+                dependency.manifest.version
+            );
```
Comment (wngr) [inline src/package.rs:135]: If there's concurrent stuff happening the output might be intertwined. Maybe write it all at once?

File: src/package.rs [hunk 2]
```diff
@@ -159,32 +250,94 @@ impl PackageStore {
...
 impl Package {
     /// Creates a new package
-    pub fn new(name: PackageId, version: String, tgz: Bytes) -> Self {
-        Self { name, version, tgz }
+    pub fn new(manifest: PackageManifest, tgz: Bytes) -> Self {
+        Self { manifest, tgz }
+    }
+
+    /// Decodes the tar and returns the encoded manifest
+    pub fn decode(tgz: Bytes) -> eyre::Result<Self> {
```
Comment (wngr) [inline src/package.rs:276]: `impl TryFrom<Bytes>`?

**Review by wngr — COMMENTED** [review]
Isolated changes LGTM. Could I trouble you for some tests :-)?

**Comment by mara214** [comment]
> Isolated changes LGTM. Could I trouble you for some tests :-)?

Of course!

File: src/package.rs (hunk 1 above)
Comment (mara214) [reply src/package.rs:135]: there is no concurrent stuff happening, but i agree 😊 
i tried to use tracing spans but it complains about not using them across async ops

File: src/package.rs (hunk 2 above)
Comment (mara214) [reply src/package.rs:276]: true!

but the weird thing here is that we only do some introspection and then still store the compressed tgz

But i agree when seeing this a second time, it should be try from

File: src/package.rs (hunk 1 above)
Comment (wngr) [reply src/package.rs:135]: I think you can annotate the whole function or use a closure, see https://docs.rs/tracing/latest/tracing/struct.Span.html#in-asynchronous-code.

However this won't help here, as individual log events will be pushed to your subscriber anyway, just potentially augmented with some span metadata.
Spans might be interesting nonetheless get some timing information 🤷

Comment (mara214) [reply src/package.rs:135]: Okay :) I'll fix the logging! Thx for the hint

**Pull request merged**